### PR TITLE
Fix duplicate connection

### LINKED ISSUE #1123: When one side connect actively, the other side will always keep trying a new connection
**Describe the bug**

If neo node A and B both listen port 40333, and A connected B firstly. At this time, B will always try to create a new connection from B to A with timer.
But when the new connection created, and B sent the `VersionPlayload` to A, A will close the connection at once, as A have connected B. 

https://github.com/neo-project/neo/blob/c4023a48567eea8a82967ad91e436c42b5552511/neo/Network/P2P/RemoteNode.cs#L211-L216

> Assumption: The number of connected nodes is less than `MinDesiredConnections`.

**To Reproduce**

Evn.:  4 four nodes in different hosts.

1. add log in RemoteNode

```csharp
 public RemoteNode(NeoSystem system, object connection, IPEndPoint remote, IPEndPoint local)
            : base(connection, remote, local)
        {
            System.Console.WriteLine("add RemoteNode " + remote.ToString() + "  remotes: " + LocalNode.Singleton.RemoteNodes.Count);  // <---- here

            this.system = system;
            this.protocol = Context.ActorOf(ProtocolHandler.Props(system));
            LocalNode.Singleton.RemoteNodes.TryAdd(Self, this);
         ...............................
```

 add log in RemoteNode.PostStop

```csharp

        protected override void PostStop()
        {
            LocalNode.Singleton.RemoteNodes.TryRemove(Self, out _);
            base.PostStop();

            System.Console.WriteLine("remove RemoteNode " + Remote.ToString() + "  remotes: " + LocalNode.Singleton.RemoteNodes.Count);  // <------- 
        }
```

 add log in Peer.OnTimer

```csharp
 private void OnTimer()
        {
            if (ConnectedPeerActors.Count >= MinDesiredConnections) return;
            if (UnconnectedPeers.Count == 0)
                NeedMorePeers(MinDesiredConnections - ConnectedPeerActors.Count);
            foreach(IPEndPoint unconnected in UnconnectedPeers)
            {
                Console.WriteLine("--- unconnected " + unconnected.ToString());  // <--- here
            }
            foreach (IPEndPoint connected in ConnectedPeers.Values)
            {
                Console.WriteLine("--- connected " + connected.ToString());   // <--- here
            }

            IPEndPoint[] endpoints = UnconnectedPeers.Take(MinDesiredConnections - ConnectedPeerActors.Count).ToArray();
            ImmutableInterlocked.Update(ref UnconnectedPeers, p => p.Except(endpoints));
            foreach (IPEndPoint endpoint in endpoints.Except(ConnectedPeers))
            {
                Console.WriteLine("start to connect " + endpoint.ToString() + " ..."); // <--- here
                ConnectToPeer(endpoint);
            }
        }
```

2. setup a privatenet and watch the logs

**Expected behavior**

When A connected B, B dose not need to create a new connection between B and A unless the original connection crashed.

**Screenshots**

image

B's ip is 192.168.1.147, and A's ip is 192.168.1.149.  A connected B firstly. As you can see, B will try to create a connection with timer.

**Platform:**
 - OS: Centos 7.0
 - Version neo-cli 3.0.0 preview1

**How to solve?**

The reason is that A dosen't use the listener port 40333 but other port to connect B, but B will treat the `{A's ip}:40333` as a new node, and try to connect it.

When A connect B, B must also add the `{A's ip}:{A's listener port}` in its connected peers.

### PR DESCRIPTION
**Pull request by Tommo-L**
Close #1123 

- filter duplicate connections automatically

As duplicate connection will affect tps testing, I want to send this single pr.

### CODE REVIEW

**Comment by Tommo-L** [comment]
Test result:
Node1, node2, node3 all will connect each other one time without duplicate connection, after connected.

image

File: src/neo/Network/P2P/LocalNode.cs [hunk 1]
```diff
@@ -127,6 +127,35 @@ internal static IPEndPoint GetIpEndPoint(string hostAndPort)
...
+        /// If it is equal to the Nonce of other RemoteNode, we just return true, else we'll return false and update the Listener address of the connected remote node.
+        /// </summary>
+        /// <param name="remoteActor">Remote node actor</param>
+        /// <param name="remoteNode">Remote node</param>
+        public bool CheckDuplicateNonce(IActorRef remoteActor, RemoteNode remoteNode)
+        {
+            var version = remoteNode.Version;
+            var remote = remoteNode.Remote;
+
+            if (remote is null) return false;
```
Comment (shargon) [inline src/neo/Network/P2P/LocalNode.cs:141]: It's better to rename the method, something like `AllowNewConnection` and also disconnect in this condition

File: src/neo/Network/P2P/RemoteNode.ProtocolHandler.cs [hunk 2]
```diff
@@ -356,12 +356,7 @@ private void OnVersionMessageReceived(VersionPayload payload)
                         break;
                 }
             }
-            if (payload.Nonce == LocalNode.Nonce || payload.Magic != ProtocolSettings.Default.Magic)
-            {
-                Disconnect(true);
-                return;
-            }
-            if (LocalNode.Singleton.CheckDuplicateNonce(Self, this))
+            if (payload.Magic != ProtocolSettings.Default.Magic || LocalNode.Singleton.AllowNewConnection(Self, this))
```
Comment (shargon) [inline src/neo/Network/P2P/RemoteNode.ProtocolHandler.cs]:
```suggestion
            if (payload.Magic != ProtocolSettings.Default.Magic || !LocalNode.Singleton.AllowNewConnection(Self, this))
```

File: src/neo/Network/P2P/LocalNode.cs [hunk 3]
```diff
@@ -127,6 +127,35 @@ internal static IPEndPoint GetIpEndPoint(string hostAndPort)
...
+        /// If it is equal to the Nonce of local or any remote node, it'll return false, else we'll return true and update the Listener address of the connected remote node.
+        /// </summary>
+        /// <param name="remoteActor">Remote node actor</param>
+        /// <param name="remoteNode">Remote node</param>
+        public bool AllowNewConnection(IActorRef remoteActor, RemoteNode remoteNode)
+        {
+            var version = remoteNode.Version;
+            var remote = remoteNode.Remote;
+
+            if (remote is null) return true;
```
Comment (shargon) [inline src/neo/Network/P2P/LocalNode.cs:141]: false?
```suggestion
            if (remote is null) return false;
```

Comment (Tommo-L) [reply src/neo/Network/P2P/LocalNode.cs:141]: Okay, some ut will pass remote = null, I'll change it.

File: src/neo/Network/P2P/LocalNode.cs [hunk 4]
```diff
@@ -127,6 +127,30 @@ internal static IPEndPoint GetIpEndPoint(string hostAndPort)
...
 
+        /// <summary>
+        /// Check the new connection <br/>
+        /// If it is equal to the Nonce of local or any remote node, it'll return false, else we'll return true and update the Listener address of the connected remote node.
+        /// </summary>
+        /// <param name="actor">Remote node actor</param>
+        /// <param name="node">Remote node</param>
+        public bool AllowNewConnection(IActorRef actor, RemoteNode node)
+        {
+            if (node.Remote is null) return false;
```
Comment (erikzhang) [inline src/neo/Network/P2P/LocalNode.cs:138]: This cannot be `null` because it is assigned in the constructor.

Comment (Tommo-L) [reply src/neo/Network/P2P/LocalNode.cs:138]: Yes, it can't be null now.

**Pull request merged**